### PR TITLE
directory-mode: sort by name and mtime

### PR DESCRIPTION
**Pull request by vindarel**
It works :) and it can certainly be improved (before being extended).

### CODE REVIEW

**Comment by cxxxr** [comment]
This is a very good improvement.
Thank you very much!

Does "draft" mean there will be more changes to this PR?

**Comment by vindarel** [comment]
Thank you!

Yes, I put [draft] because I plan on improving the PR a little bit. I opened the PR to collect early feedback.

Note that I used a `+sbcl` feature flag, and I threw an error for other implementations… (I didn't want to pull the osicat library as a dependency).

File: src/base/buffer.lisp [hunk 1]
```diff
@@ -82,7 +82,11 @@
     :accessor buffer-encoding)
    (last-write-date
     :initform nil
-    :accessor buffer-last-write-date)))
+    :accessor buffer-last-write-date)
+   (keys
+    :initform (make-hash-table)
+    :accessor buffer-keys
+    :documentation "Store internal keys and values in a hash-table. Example: store the current sorting method of a directory buffer.")))
```
Comment (cxxxr) [inline src/base/buffer.lisp:89]: There is already a slot called buffer-variables, can this be used?

e.g.
```
(buffer-value buffer '<key>) => <value>
(setf (buffer-value buffer '<key>) <value>) => set <value>
```

File: src/base/file-utils.lisp [hunk 2]
```diff
@@ -65,21 +65,35 @@
...
+
 (defun directory-files (pathspec)
   (if (uiop:directory-pathname-p pathspec)
       (list (pathname pathspec))
       (or (mapcar (lambda (x) (virtual-probe-file x pathspec))
                   (directory pathspec))
           (list pathspec))))
 
-(defun list-directory (directory &key directory-only)
+(defun list-directory (directory &key directory-only (sort-method 'pathname))
```
Comment (cxxxr) [inline src/base/file-utils.lisp:88]: Can we define sort-method as a type?

```
(deftype sort-method ()
  '(member :pathname :mtime))
```

File: src/ext/directory-mode.lisp [hunk 3]
```diff
@@ -1,9 +1,16 @@
 (defpackage :lem/directory-mode
   (:use :cl :lem)
   #+sbcl
-  (:lock t))
+  (:lock t)
+  (:export
+   #:*default-sort-method*))
```
Comment (cxxxr) [inline src/ext/directory-mode.lisp:6]: For consistency, it might be a good idea to use the :keyword

**Review by cxxxr — COMMENTED** [review]
Thanks for the nice PR!
I have made a few comments.

**Comment by vindarel** [comment]
Nice, thanks. Should be all good! 

(I didn't rebase and force-push, I'll let you rebase while merging if you want)

File: src/base/package.lisp [hunk 4]
```diff
@@ -111,6 +111,7 @@
    :buffer-undo-boundary
    :buffer-value
    :buffer-unbound
+   :buffer-keys
```
Comment (cxxxr) [inline src/base/package.lisp:114]: It appears that this export remains

**Review by cxxxr — COMMENTED** [review]
I think it's very good.
A minor point, but I commented on it.
I am not very particular about rebase.
Thank you!

File: src/base/file-utils.lisp (hunk 2 above)
[the remark this answers is not in the document]
Comment (vindarel) [reply src/base/file-utils.lisp:88]: It was working because it was handled by the "otherwise" clause of `cond`, but good catch.

**Comment by cxxxr** [comment]
Thank you so much!

**Pull request merged**